vertical-pod-autoscaler/pkg/crdcache/crdcache.go

```diff
@@ -0,0 +1,161 @@
+	c.extensionLister = extensionInformer.Lister()
+	return c
+}
+
+type crdCache struct {
+	dynamicFactory  dynamicinformer.DynamicSharedInformerFactory
+	extensionLister extensionlister.CustomResourceDefinitionLister
+	// startedInformers caches started GroupVersionResource informers
+	startedInformers map[schema.GroupVersionResource]bool
+	lock             sync.Mutex
+}
+
+func (c *crdCache) FetchSelector(gvr schema.GroupVersionResource, namespace, name string) (labels.Selector, error) {
+	crd, err := c.extensionLister.Get(gvr.GroupResource().String())
+	if err != nil {
+		return nil, err
+	}
+	subRes := crd.Spec.Subresources
+	if subRes == nil || subRes.Scale == nil || subRes.Scale.LabelSelectorPath == nil ||
+		*subRes.Scale.LabelSelectorPath == "" {
+		return nil, fmt.Errorf("resource %s hasn't scale sub-resource", gvr.String())
+	}
+	lister := c.getLister(gvr)
+	obj, err := lister.ByNamespace(namespace).Get(name)
+	if err != nil {
+		return nil, err
+	}
+	unstructuredObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return nil, fmt.Errorf("crd object %s %s/%s can't be converted to a unstructured object", gvr.String(), name, namespace)
+	}
+	selectorStr, err := selectorFromCustomResource(unstructuredObj, *subRes.Scale.LabelSelectorPath)
+	if err != nil {
+		return nil, fmt.Errorf("can't get label selector from crd object %s %s/%s: %v", gvr.String(), name, namespace, err)
+	}
+	selector, err := labels.Parse(selectorStr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse label selector %s from crd object %s %s/%s: %v", selectorStr, gvr.String(), name, namespace, err)
+	}
+	return selector, nil
+}
+
+func (c *crdCache) IsScalable(gvr schema.GroupVersionResource, namespace, name string) (bool, error) {
+	crd, err := c.extensionLister.Get(gvr.GroupResource().String())
+	if err != nil {
+		return false, err
+	}
+	subRes := crd.Spec.Subresources
+	if subRes == nil || subRes.Scale == nil {
+		return false, nil
+	}
+	lister := c.getLister(gvr)
+	_, err = lister.ByNamespace(namespace).Get(name)
+	return err == nil, err
+}
+
+func (c *crdCache) GetOwnerReferences(gvr schema.GroupVersionResource, namespace, name string) ([]metav1.OwnerReference, error) {
+	lister := c.getLister(gvr)
+	obj, err := lister.ByNamespace(namespace).Get(name)
+	if err != nil {
+		return nil, err
+	}
+	unstructuredObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return nil, fmt.Errorf("crd object %s %s/%s can't be converted to a unstructured object", gvr.String(), name, namespace)
+	}
+	return unstructuredObj.GetOwnerReferences(), nil
+}
+
+// getLister gets the lister for the given GroupVersionResource and makes sure the informer have started and synced
+func (c *crdCache) getLister(gvr schema.GroupVersionResource) cache.GenericLister {
+	informer := c.dynamicFactory.ForResource(gvr)
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if _, ok := c.startedInformers[gvr]; !ok {
+		klog.Infof("watching custom resource definition %s", gvr.String())
```

Comment:
nit: logs start with capital letters.

- fmt.Error: lower case letter - error will be concatenated.
- klog.Info: capital letter - responsible for printing